packages/nitro-protocol/contracts/NitroAdjudicator.sol

```diff
@@ -2,87 +2,17 @@
 pragma solidity 0.7.4;
 pragma experimental ABIEncoderV2;
 
-import './interfaces/IAdjudicator.sol';
 import './ForceMove.sol';
 import './Outcome.sol';
-import './AssetHolder.sol';
+import './MultiAssetHolder.sol';
 
 /**
- * @dev The NitroAdjudicator contract extends ForceMove and hence inherits all ForceMove methods, and also extends and implements the Adjudicator interface, allowing for a finalized outcome to be pushed to an asset holder.
+ * @dev The NitroAdjudicator contract extends MultiAssetHolder and hence inherits all MultiAssetHolder methods.
  */
-contract NitroAdjudicator is IAdjudicator, ForceMove {
+contract NitroAdjudicator is MultiAssetHolder {
```

Comment:
I agree that the existing separation of concerns "seems nice"; but what is it precisely that we like about that separation of responsibility? The whole point of this PR is to remove the separation of responsibilities for managing a channel -- from many contracts to one contract. And even further than that, from many storage slots to a single storage slot. 

Your suggestion for `initiateExit` is somewhat reminiscent of `pushOutcome`, although it retains the (gas) benefits of only using a single storage slot. A method that essentially ends the adjudication phase of the channel and begins the exit phase. Such a method allows for separation of responsibilities in time, rather than separation in "space" (memory). I like the idea, but I'm not sure I see a concrete benefit. And I do see an increased complexity (introduction of a new method that I don't think we strictly need) and the associated gas increase. 

It is somehow dissatisfying that we need to provide non-asset related data to the "exit" methods `transfer` and `claim`. But if I understand your proposal correctly, the `initiateExit` method would require that data, and would need to be applied at least once before those exit methods. So the benefit seems quite small, to me. 

Side note on contract inheritance: I had originally wanted to have `NitroAdjudicator is ForceMove, MultiAssetHolder` inheritance relationship, but found this difficult to get working. It might still be possible though, and would avoid the need for `MultiAssetHolder` to inherit `ForceMove`. Alternatively, I don't have an issue with pushing everything from `NitroAdjudicator` into `MultiAssetHolder` and renaming it to `NitroAdjudicator`.

Reply:
> I agree that the existing separation of concerns "seems nice"; but what is it precisely that we like about that separation of responsibility? The whole point of this PR is to remove the separation of responsibilities for managing a channel -- from many contracts to one contract. And even further than that, from many storage slots to a single storage slot.

There is a significant benefit in terms of understanding the code. For instance, to understand how a `MultiAssetHolder` depends on being a `ForceMove`, I'd need to
1. enumerate the functions available from` ForceMove`
2. check which of those functions are called by `MultiAssetHolder`

> Your suggestion for `initiateExit` is somewhat reminiscent of `pushOutcome`, although it retains the (gas) benefits of only using a single storage slot. A method that essentially ends the adjudication phase of the channel and begins the exit phase. Such a method allows for separation of responsibilities in time, rather than separation in "space" (memory) ... I like the idea, but I'm not sure I see a concrete benefit.

A clear split in time is indeed a very important property, so making the split in time obvious is one important benefit!

>  And I do see an increased complexity (introduction of a new method that I don't think we strictly need) and the associated gas increase.

> It is somehow dissatisfying that we need to provide non-asset related data to the "exit" methods `transfer` and `claim`. But if I understand your proposal correctly, the initiateExit method would require that data, and would need to be applied at least once before those exit methods. So the benefit seems quite small, to me.

As a comment, `challengerAddress` does not appear to be needed in the fingerprint. Removing `challengerAddress` as well as `stateHash` would provide a small gas savings, and simplify the Nitro API. (See this [draft PR](https://github.com/statechannels/statechannels/pull/3684))

Furthermore, while `initiateExit` is nice, it is not optimized. You can easily get the same effect by keeping the state hash in the asset holder API, and:
- when the provided state hash is `bytes32(0)`, check that the value of `statusOf` is equal to the outcome hash.  otherwise, check that the fingerprint matches
- when the asset manager updates `statusof`, always record the `outcomeHash`


> Side note on contract inheritance: I had originally wanted to have `NitroAdjudicator is ForceMove, MultiAssetHolder` inheritance relationship, but found this difficult to get working. It might still be possible though, and would avoid the need for `MultiAssetHolder` to inherit `ForceMove`.

I don't know if Solidity supports this, but I feel like it should:
- Define a `StatusManager` contract
- `MultiAssetHolder` inherits from `StatusManager`
- `ForceMove` inherits from `StatusManager`
- `NitroAdjudicator` inherits from `MultiAssetHolder` and `ForceMove`

Note that `MultiAssetHolder` _only needs_ `requireChannelFinalized`, and `ForceMove` _does not need_ `requireChannelFinalized`, so this architecture also provides a well-defined "separation of responsibility", without needing to perform the `statusOf` ninja fu described above.
- `MultiAssetHolder` manages `statusOf` when the channel is finalized
- `ForceMove` manages `statusOf` when the channel is not finalized.

Reply:
https://github.com/statechannels/statechannels/issues/3685